### PR TITLE
feat(Thread): dispatch `on_thread_create` event and add `_newly_created`

### LINKED ISSUE #323: Track thread.newly_created
### Summary

implement api info to more easily tell thread events apart

### What is the feature request for?

The core library

### The Problem

Its not possible to tell exactly what the thread event is at the moment, whether the bot was added to the thread or the thread was just created.

### The Ideal Solution

Implement https://github.com/discord/discord-api-docs/commit/1875d80011ea62bf555ce807459ccf97950291e9

### The Current Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Victorsitou**
## Summary

1. Added `Thread._newly_created` (it's private since in the other events, such as `on_thread_join` etc, it gets the cached Thread, so having that attribute public would be inaccurate with the actual behavior)
2. Added `on_thread_create` event, and deleted handling of `on_thread_join` in `parse_thread_create` since Discord seems to call `THREAD_MEMBERS_UPDATE` (for when a user joins a thread) and `THREAD_CREATE` (for the same purpose), so it was dispatching `on_thread_join` twice (pretty sure it was an issue before too)
<sup>A little more context can be found [here](https://canary.discord.com/channels/808030843078836254/937828640463401001/939658154613039175)</sup>

referece:
https://github.com/discord/discord-api-docs/commit/1875d80011ea62bf555ce807459ccf97950291e9

Closes #323

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint` or `pre-commit run --all-files`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/state.py [hunk 1]
```diff
@@ -1034,11 +1034,15 @@ def parse_thread_create(self, data) -> None:
             _log.debug("THREAD_CREATE referencing an unknown guild ID: %s. Discarding", guild_id)
             return
 
+        if data.get("newly_created") is None:
```
Comment (EQUENOS) [inline disnake/state.py:1037]: `if "newly_created" not in data:` is probably a better way

File: disnake/threads.py [hunk 2]
```diff
@@ -182,6 +183,7 @@ def _from_data(self, data: ThreadPayload):
         self.slowmode_delay = data.get("rate_limit_per_user", 0)
         self.message_count = data.get("message_count")
         self.member_count = data.get("member_count")
+        self._newly_created: Optional[bool] = data.get("newly_created")
```
Comment (EQUENOS) [inline disnake/threads.py:186]: Defaults to `False`?

Comment (Victorsitou) [reply disnake/threads.py:186]: I don't think it's necessary, it's a private attribute anyway.

File: disnake/state.py (hunk 1 above)
Comment (Victorsitou) [reply disnake/state.py]: changed in c6a8925d12b80e3419dc85f26f16d40df57fb4e1, thanks

**Review by shiftinv — COMMENTED** [review]
nitpicks

File: docs/api.rst [hunk 3]
```diff
@@ -762,8 +762,12 @@ to handle it, which defaults to print a traceback and ignoring the exception.
...
 .. function:: on_thread_join(thread)
 
-    Called whenever a thread is joined or created. Note that from the API's perspective there is no way to
-    differentiate between a thread being created or the bot joining a thread.
+    Called whenever your bot joins a thread.
+
+    .. versionchanged:: 2.4
+
+        This is now only dispatched when the bot joins a thread, not when a thread is created.
+        If you want an event for when a thread is created, use :ref:`on_thread_created` instead.
```
[the remark this answers is not in the document]
Comment (Victorsitou) [reply docs/api.rst:770]: 😔

**Review by shiftinv — CHANGES_REQUESTED** [review]
ftr, this currently doesn't quite work as expected with private threads

https://discord.com/channels/808030843078836254/932405291314729030/941730937354465290
https://discord.com/channels/808030843078836254/941736646880165929

**Review by EQUENOS — APPROVED** [review]
lgtm

**Review by EQUENOS — CHANGES_REQUESTED** [review]
Nvm, found a bug when invited the bot to another server

File: disnake/threads.py (hunk 2 above)
Comment (onerandomusername) [inline disnake/threads.py:186]: why is this saved at all? We can dispatch a on_thread_create if its true in the received data, in which case the thread will be known to be newly created and we do not need to save it.

File: disnake/state.py [hunk 4]
```diff
@@ -1222,6 +1252,23 @@ def parse_guild_member_update(self, data) -> None:
...
+                old_perm_value = thread.permissions_for(old_member).value
+                new_perm_value = thread.permissions_for(member).value
+                # get bits that were flipped from 1 to 0
+                diff_perm_value = old_perm_value ^ new_perm_value & old_perm_value
+                if (
+                    diff_perm_value
+                    and Permissions(diff_perm_value).view_channel
+                    and member.id in thread._members
+                ):
+                    self.dispatch("thread_remove", thread)
```
Comment (onerandomusername) [inline disnake/state.py:1270]: Is this technically accurate? The thread still exists, and the bot *is* still a member, even if the bot cannot see it.

File: docs/api.rst [hunk 5]
```diff
@@ -762,8 +762,12 @@ to handle it, which defaults to print a traceback and ignoring the exception.
 
 .. function:: on_thread_join(thread)
 
-    Called whenever a thread is joined or created. Note that from the API's perspective there is no way to
-    differentiate between a thread being created or the bot joining a thread.
+    Called whenever your bot joins a thread.
+
+    .. versionchanged:: 2.4
```
Comment (onerandomusername) [inline docs/api.rst:767]: 2.5

File: docs/api.rst [hunk 6]
```diff
@@ -794,6 +798,19 @@ to handle it, which defaults to print a traceback and ignoring the exception.
...
 
+.. function:: on_thread_create(thread)
+
+    Called whenever a thread is created.
+
+    Note that you can get the guild from :attr:`Thread.guild`.
+
+    This requires :attr:`Intents.guilds` to be enabled.
+
+    .. versionadded:: 2.4
```
Comment (onerandomusername) [inline docs/api.rst:809]: 2.5

File: disnake/state.py [hunk 7]
```diff
@@ -1034,10 +1053,19 @@ def parse_thread_create(self, data) -> None:
             _log.debug("THREAD_CREATE referencing an unknown guild ID: %s. Discarding", guild_id)
             return
 
+        if int(data["type"]) == 12 and "member" in data:
+            # We don't want to create `Thread` instances and dispatch `on_thread_join` for when
+            # the bot joins a public thread. As this is handled in `parse_thread_members_update`.
+            return
```
Comment (onerandomusername) [inline disnake/state.py:1059]: news threads are also public but have a different channel type

**Review by onerandomusername — CHANGES_REQUESTED** [review]
lots of things to look at

File: disnake/threads.py (hunk 2 above)
Comment (EQUENOS) [reply disnake/threads.py:186]: Because it's technically accurate

File: disnake/state.py (hunk 4 above)
Comment (EQUENOS) [reply disnake/state.py:1270]: It's not. However, there's no other way to know if you're removed from a private thread or not. If we don't dispatch thread_remove here and someone removes the bot afterwards, we won't dispatch it at all.

File: disnake/threads.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/threads.py:186]: But we don't need it. It is used in one function, and is only True in that specific gateway event, it should not be stored.

Discord's documentation does not document it as a thread attribute, there is absolutely no reason to store it as such.

**Comment by shiftinv** [comment]
After some discussion over the last few weeks we've decided to reapproach this in separate PRs, as this PR eventually kind of suffered from scope creep and the addition of the `newly_created` field is more important than taking care of minor cache eviction issues for now.

**Pull request closed without merging**